engine/logger.h

```diff
@@ -0,0 +1,150 @@
+        temp = "[{}] " + temp;
+        format = temp.c_str();
+
+        const int logNum = Logger::GetNextLogNumber();
+        switch (level)
+        {
+        case spdlog::level::trace:
+            spdlog::trace(format, logNum, args...);
+            break;
+        case spdlog::level::debug:
+            spdlog::debug(format, logNum, args...);
+            break;
+        case spdlog::level::info:
+            spdlog::info(format, logNum, args...);
+            break;
+        case spdlog::level::warn:
+            spdlog::warn(format, logNum, args...);
+            break;
+        case spdlog::level::err:
+            spdlog::error(format, logNum, args...);
+            break;
+        case spdlog::level::critical:
+            spdlog::critical(format, logNum, args...);
+            break;
+        default:
+            break;
+        }
+    };
+
+private:
+    inline static int sLogNumber { 0 };
+};
+
+inline std::unique_ptr<Logger> gLogger { nullptr };
+inline std::unique_ptr<RealTimeLogger> gRTLogger { nullptr };
+
+template<typename... Args>
+void LogDebug(const char* format, Args... args)
+{
+    if (gLogger != nullptr)
+    {
+        gLogger->Log(spdlog::level::debug, format, args...);
+    }
+}
+
+template<typename... Args>
+void LogInfo(const char* format, Args... args)
+{
+    if (gLogger != nullptr)
+    {
+        gLogger->Log(spdlog::level::info, format, args...);
+    }
+}
+
+template<typename... Args>
+void LogWarning(const char* format, Args... args)
+{
+    if (gLogger != nullptr)
+    {
+        gLogger->Log(spdlog::level::warn, format, args...);
+    }
+}
+
+template<typename... Args>
+void LogCritical(const char* format, Args... args)
+{
+    if (gLogger != nullptr)
+    {
+        gLogger->Log(spdlog::level::critical, format, args...);
+    }
+}
+
+template<typename... Args>
+void LogError(const char* format, Args... args)
+{
+    if (gLogger != nullptr)
+    {
+        gLogger->Log(spdlog::level::err, format, args...);
+    }
+}
+
+template<typename... Args>
+void RealTimeLogDebug(const char* format, Args... args)
+{
+    if (gRTLogger != nullptr)
+    {
+        gRTLogger->Log(spdlog::level::debug, format, args...);
+    }
+}
+
+template<typename... Args>
+void RealTimeLogInfo(const char* format, Args... args)
+{
+    if (gRTLogger != nullptr)
+    {
+        gRTLogger->Log(spdlog::level::info, format, args...);
+    }
+}
+
+template<typename... Args>
+void RealTimeLogWarning(const char* format, Args... args)
+{
+    if (gRTLogger != nullptr)
+    {
+        gRTLogger->Log(spdlog::level::warn, format, args...);
+    }
+}
+
+template<typename... Args>
+void RealTimeLogCritical(const char* format, Args... args)
+{
+    if (gRTLogger != nullptr)
+    {
+        gRTLogger->Log(spdlog::level::critical, format, args...);
+    }
+}
+
+template<typename... Args>
+void RealTimeLogError(const char* format, Args... args)
+{
+    if (gRTLogger != nullptr)
+    {
+        gRTLogger->Log(spdlog::level::err, format, args...);
```

Comment:
you may want to (for all these calls) do std::forward for the variadic args.

This is going to get really into the weeds of nasty template BS, but I think the purely correct version of something like this would be:

```


#include <utility> // for std::forward

template<typename... Args>
void RealTimeLogCritical(const char* format, Args&&... args)
{
    if (gRTLogger != nullptr)
    {
        gRTLogger->Log(spdlog::level::critical, format, std::forward<Args>(args)...);
    }
}

```

The very short explanation is this will preserve the "value category" of everything in the argument pack, that is rvalue references, lvalue references, const lvalue references, const rvalue references ........... will all remain the same way they are passed in, instead of being converted here.

There was a great talk at this year's cppcon on forwarding references that I recommend if you want to grok this.

I would not spend a ton of time wrapping your head around this unless you're really curious. I would recommend:

1. ignoring me, it doesn't really matter
2. trusting me, and just using it

either option totally fine, just wanted to bring it up

Reply:
I'll trust ya, you've been right before once or twice 😁